### PR TITLE
Remove ref struct pinning from locals

### PR DESCRIPTION
**Pull request by MichalPetryka**
Removes pinned from `ref struct` locals as refs to them can't point to the heap legally.

### CODE REVIEW

File: src/coreclr/jit/lclvars.cpp [hunk 1]
```diff
@@ -275,7 +275,26 @@ void Compiler::lvaInitTypeRef()
...
+                {
+                    refClsHnd = realClsHnd;
+                }
+            }
+
+            // ignore pinned on ref structs
+            if ((refClsHnd != NO_CLASS_HANDLE) &&
+                ((info.compCompHnd->getClassAttribs(refClsHnd) & CORINFO_FLG_BYREF_LIKE) != 0))
+            {
+                JITDUMP("Ignoring pinned on local V%02u for byref-like type %s\n", varNum, eeGetClassName(refClsHnd));
```
Comment (MichalPetryka) [inline src/coreclr/jit/lclvars.cpp:295]: @jkotas We were having a discussion on Discord whether enforcing `ref struct` `ref`s not pointing to the heap is an acceptable assumption for the JIT to make, could you give your opinion on the safety of such assumption?

Comment (jakobbotsch) [reply src/coreclr/jit/lclvars.cpp:295]: The concern is about aliasing. Before this change the JIT's aliasing model allows roundtripping any `ref A` as a `ref B`. The optimization in this PR is not valid unless we consider that to be UB when A and B do not match in byref likeness.

**Comment by VSadov** [comment]
Just curious - what are the savings here? Pinning references that happen to be stack refs seems like mostly a noop - basically just a different bit in GCinfo with no effect on GC. Once GC figures that byref does not point to the heap (which it needs to figure regardless), it will not care about pinning. Actually will stop caring about the reference entirely.

Are there some possible savings on the JIT side?

**Comment by MichalPetryka** [comment]
> Just curious - what are the savings here? Pinning references that happen to be stack refs seems like mostly a noop - basically just a different bit in GCinfo with no effect on GC. Once GC figures that byref does not point to the heap (which it needs to figure regardless), it will not care about pinning. Actually will stop caring about the reference entirely.
> 
> Are there some possible savings on the JIT side?

The motivation was to improve the codegen a bit as some people use such pinning to workaround the lack of `Unsafe` for ref structs.

Example:
image
Before:
image
After:
image

**Comment by VSadov** [comment]
> The motivation was to improve the codegen a bit

I see. That is basically because pinned locals cannot be enregistered, so pinning will force it to the stack.  
I kind of suspected that, just was not sure if it can cause a real grief to people or perhaps it is something else.

**Comment by Sergio0694** [comment]
Does this also ignores pinning for fields of ref struct types (pointed to by ref)? Eg.:
```csharp
fixed (long* p = &r2.a)
{
}
```
(using the same types and names as that snippet in https://github.com/dotnet/runtime/pull/97997#issuecomment-1967580647)

**Comment by MichalPetryka** [comment]
> Does this also ignores pinning for fields of ref struct types (pointed to by ref)? Eg.:
> 
> ```cs
> fixed (long* p = &r2.a)
> {
> }
> ```
> 
> (using the same types and names as that snippet in [#97997 (comment)](https://github.com/dotnet/runtime/pull/97997#issuecomment-1967580647))

No as the local is not a ref to a ref struct in such case.

**Comment by Sergio0694** [comment]
Not sure I'm following. What I meant is, if you have:

```csharp
ref struct MyRefStruct
{
    public int X;
}

void Test(ref MyRefStruct value)
{
    fixed (int* p = &value.X)
    {
        DoStuff(p);
    }
}
```

Shouldn't it also be possible to ignore the pinning on that `X` field? 🤔

**Comment by MichalPetryka** [comment]
> Not sure I'm following. What I meant is, if you have:
> 
> ```cs
> ref struct MyRefStruct
> {
>     public int X;
> }
> 
> void Test(ref MyRefStruct value)
> {
>     fixed (int* p = &value.X)
>     {
>         DoStuff(p);
>     }
> }
> ```
> 
> Shouldn't it also be possible to ignore the pinning on that `X` field? 🤔

The optimization here is done before the JIT even looks at the IL opcodes so it doesn't know what's assigned to the local, the case you've mentioned would need to be handled later on by looking at the struct layout the `FIELD_ADDR` for X is coming from and checking if the type there is a ref struct. I think it should be possible but I'm not 100% sure if "unpinning" the local would have any issues then.

**Comment by jakobbotsch** [comment]
Unfortunately I don't think the impact of the optimization warrants us changing the model of what kind of byref aliasing we consider legal within the JIT. If we could demonstrate some larger impact, then I wouldn't be against reconsidering. Thanks!

**Pull request closed without merging**